### PR TITLE
Change GET events endpoints

### PR DESCRIPTION
**Pull request by mattlub**
**UPDATE 14/8/17: READY FOR REVIEW**

This has turned into a bigger PR than intended, but not too much deep stuff going on.

Includes 
- [x] populate placeId field with place information for getAll and getById
- [x] sort events by date in getAll
- [x] convert controllers to use new `rejectIfNull` function, remove old custom schema methods
- [x] update docs

doesn't include (can come in another PR):
- [ ] change `placeId` field to `place` (?) #96 
- [ ] date-related query params for events

-----

after discussing, I think these need to be changed, as Mohamed is working on the nazareth-events and it is already causing problems.

relates #81 #89 

notes (outdated):
- brings up need for dropCollection**s**AndEnd
- cannot implement population easily for getById because of the helper functions in db Utils
- docs will need to be updated

### CODE REVIEW

**Comment by mattlub** [comment]
@des-des @m4v15 

I updated the getAll function [as so](https://github.com/foundersandcoders/open-tourism-platform/blob/f5b3b2274e707116826d16cc048846d28a0c6f86/src/controllers/event.js#L7-L12) and it works nicely.

However, I cannot add `.populate` to the custom `getByIdOrError` function (defined [here](https://github.com/foundersandcoders/open-tourism-platform/blob/f5b3b2274e707116826d16cc048846d28a0c6f86/src/db/utils.js#L10-L18)). I kind of understand why it shouldn't, it's no longer a mongoose `query` so you can't do `.populate`, but I'm not certain about lots of promises stuff.

Is there any solution you can see, the only ideas I have are:
- pass an options object to the `getByIdOrError` function, including populate options (quite messy)
- somehow construct and return a mongoose query instead of a resolved/rejected promise (difficult)

**Comment by m4v15** [comment]
@mattlub When you say you can't add it to the getByIdOrError function, do you mean you can't chain .populate after it, or you can't literally put it inside where we define the function, at the end of [this line](https://github.com/foundersandcoders/open-tourism-platform/blob/f5b3b2274e707116826d16cc048846d28a0c6f86/src/db/utils.js#L11)?

**Comment by mattlub** [comment]
@m4v15 can't chain it on `getByIdOrError`. Don't want to put it inside because the function is generic and used for all the models.

**Comment by m4v15** [comment]
Ah I see - I think passing a populate options then wouldn't be super messy eg:
```
const findByIdOrError = function (id, fieldsToPopulate) {
  return this.findById(id)
    .populate(fieldsToPopulate||'')
    .then(res => {
      if (res === null) {
        return Promise.reject(createCustomDbError(errMessages.GET_ID_NOT_FOUND))
      }
      return Promise.resolve(res)
    })
}
```

If you pass populate an empty string it won't do anything (if you pass it null it will crash)

I mean I guess this option isn't ideal, as we're running populate when we don't need to, but I think it will work

**Comment by mattlub** [comment]
@m4v15 @des-des 
```js
const findByIdOrError = function (id, options) {
  const query = this.findById(id)
  if (options && options.populate) {
    query.populate(options.populate)
  }
  return query 
    .then(res => {
      if (res === null) {
        return Promise.reject(createCustomDbError(errMessages.GET_ID_NOT_FOUND))
      }
      return Promise.resolve(res)
    })
}
```

not overly messy, but in the controller it's slightly inconsistent because you would have:
```js
Event.find()
  .populate('place')
  .then(etc.)
```
and
```js
Event.findByIdOrError(id, {populate: 'place'})
  .then(etc.)
```

thoughts?

**Comment by m4v15** [comment]
@mattlub  Yeah or that, which is nicer (you could put the if on one line?) - or are you thinking this is too messy

**Comment by m4v15** [comment]
@mattlub, sorry I never saw your edited comment before. Yeah I mean I can't see another way of doing it really - the whole point of not doing a mongoose query was so we could get an error - the inconsistency is slightly annoying but not the end of the world?

Also I just added a dropCollectionsAndEnd function to tidy up your test - there might be a better way of doing it but recursion and promises was wigging me out a little and this way makes sense to me (somewhat).

**Comment by des-des** [comment]
I think better just to have a wrapper


```js
rejectIfEmpty = dbRes => dbRes.length === 0
  ? Promise.reject(notFoundError())
  : dbRes
}
```

```js
(req, res, next) => {
  rejectIfEmpty(Events.findOne({}))
    .then(event => res.json(res))
    .catch(next)
 }
```
Then you can handle the error in next

**Comment by des-des** [comment]
But whatever works for you

**Comment by mattlub** [comment]
@des-des  @m4v15 Have done it more like @des-des 's way, I think it's better as it removes the need for the custom mongoose model methods (`getByIdOrError` etc.). Have a look, if you approve I will implement it for all the models.

(look at the files changed instead of the individual commits because I changed things a lot in the last commit)

File: src/controllers/event.js [hunk 1]
```diff
@@ -17,9 +17,9 @@ eventController.getById = (req, res, next) => {
   // receives id in url
   // sends back one event or errors
   const id = req.params.id
-  const findEventById = Event.findById(id).populate('placeId')
-
-  errorIfNotFound(findEventById, errMessages.GET_ID_NOT_FOUND)
+  Event.findById(id)
+    .populate('placeId')
+    .then(rejectIfEmpty(errMessages.GET_ID_NOT_FOUND))
```
Comment (des-des) [inline src/controllers/event.js]: oh yum :ok_hand:

**Review by des-des — CHANGES_REQUESTED** [review]
Okay, I only have one suggestion, we call the function `rejectIfNull` as that is more accurate.

**Comment by mattlub** [comment]
@des-des @m4v15 updated and ready for review

**Review by m4v15 — APPROVED** [review]

**Review by des-des — APPROVED** [review]

**Pull request merged**